Add Tree-sitter grammar injections

maxbrunsfeld opened the pull request:
## Motivation

The TextMate parsing system allows grammars to be composed in order to syntax-highlight things like:
* JavaScript inside of script tags in an HTML file
* HTML inside of template strings in JavaScript
* SQL inside of a string literal in Python

Currently, these types of things don't work when Tree-sitter is enabled.

## Solution

This PR adds the concept of 'grammar injection' to Tree-sitter grammars. Specifically, it adds two new APIs associated with Tree-sitter grammars. These APIs might be revised before this PR is merged.

#### 1. Adding Injection Points

```js
atom.grammars.addInjectionPoint(
  languageId,
  {
    type: syntaxNodeType,
    language: languageCallback,
    content: contentCallback
  }
)
```

This method allows you to express ideas like: "In JavaScript, tagged template strings are injection points. For each tagged template string, try to identify its language by looking at the name of its tag function. Parse the content between the backticks, omitting any template substitutions."

```js
atom.grammars.addInjectionPoint(
 'javascript',
  {
    // tagged template literals are simply parsed as call expressions 
    // with a template string instead of an argument list
    type: 'call_expression', 
  
    // The language name can be found in the template string's "tag"
    language (callNode) {
      if (callNode.lastChild.type === 'template_string') {
        return callNode.firstChild.text
      }
    },

    // Parse the content inside of the template string
    content (callNode) {
      return callNode.lastChild
    }
  }
)

```
</details>


Note that this API does *not* indicate which grammar to use when parsing the content. That information can be provided by other packages, using the second API: 

#### 2. Specifying Injection Patterns

Grammars that use Tree-sitter have a new field called `injectionRegExp`. This field allows you to express ideas like: "The HTML language can be injected. Whenever there is an injection point where the language-name includes the string 'html', parse that injection point's content using the HTML parser."

```coffee
id: 'html'

injectionRegExp: 'html|HTML|Html$'
```

This two-part API allows languages to be embedded within one another without every grammar having to know about every other grammar.

## Example

[EJS](http://ejs.co) is a popular JavaScript templating system where JavaScript code is interspersed with HTML markup using the delimiters `<%` and `%>`. The HTML can of course contain more JavaScript code inside of `script` tags. And that JavaScript code can contain HTML inside of string literals.

![html-in-js-in-html-in-ejs](https://user-images.githubusercontent.com/326587/42404439-f0eecc3c-813d-11e8-9d16-084aae19525b.gif)

## Tasks

* [x] Add tests and documentation for the new `GrammarRegistry` APIs
* [x] When the buffer changes, only update the affected injections, not all of them
* [x] Update highlighting when grammars with `injectionRegExps` are added
* [x] Allow injections that are spread across many child nodes (needed for PHP, ERB, EJS, mustache, etc)
* [x] Use injections for the expand selection command
* [x] Use injections for folding specific lines
* [x] Fix syntax highlighting bugs
* [x] Use injections for getting scope descriptors
* [ ] Use injections for folding at a given nesting level

## Related Issues / PRs

Closes https://github.com/atom/atom/issues/17392
Depends on https://github.com/tree-sitter/tree-sitter/pull/181
Depends on https://github.com/tree-sitter/node-tree-sitter/pull/14
Depends on https://github.com/tree-sitter/node-tree-sitter/pull/18

maxbrunsfeld committed:
wip

maxbrunsfeld committed:
Start work on Tree-sitter grammar injections

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Get first test for grammar injections passing

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Fix syntax highlighting problems with injected languages

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Only update injections intersecting changed ranges

thomasjo reviewed (commented):
~Stupid style comment; any specific reason for prefixing some of these functions with an underscore? I'm assuming they are "private", but I don't think we do this anywhere else in the core code?~ This wasn't meant to be a review, just a line comment...

thomasjo commented on src/tree-sitter-language-mode.js:
```diff
@@ -384,11 +364,273 @@ class TreeSitterLanguageMode {
+    }
+  }
+
+  async _performUpdate (containingNode) {
```
Stupid style comment; any specific reason for prefixing some of these functions with an underscore? I'm assuming they are "private", but I don't think we do this anywhere else in the core code?

maxbrunsfeld committed:
:arrow_up: tree-sitter

maxbrunsfeld committed:
Add tests and docs for addInjectionPoint

Also, replace `addInjectionPattern` API with a single `injectionRegExp` 
field on the grammar.

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Add `text` getter to SyntaxNode

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Start parsing right away when constructing a TreeSitterLanguageMode

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Emit highlight change events when removing injections

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Define highlight iter's position in terms of tree cursor position

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Fix highlighting problem when seeking past the end of a syntax tree

maxbrunsfeld committed:
Fix possible mismatch between .findMarkers and .descendantsOfType

maxbrunsfeld committed:
Allow for parsing injections in parallel

maxbrunsfeld committed:
Update highlighting when an injectable tree-sitter grammar is loaded

Commit:
merge

maxbrunsfeld committed:
Merge branch 'master' into tree-sitter-injections

Commit:
WIP child ranges

Commit:
Merge branch 'tree-sitter-injections' of github.com:atom/atom into tree-sitter-injections

50Wliu commented:
Looks like package-lock files were committed by accident?

Commit:
injectionPoint content can return an array of nodes now

maxbrunsfeld committed:
:arrow_up: tree-sitter

maxbrunsfeld committed:
Remove package lock files

maxbrunsfeld committed:
Revert "Allow for parsing injections in parallel"

This reverts commit a484ca9e654a1da25bab08d698de42dad46e9c36.

maxbrunsfeld committed:
:arrow_up: language-javascript, language-html

maxbrunsfeld committed:
Remove grammar argument from populateInjections

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

maxbrunsfeld committed:
Fix language-javascript version

maxbrunsfeld committed:
Fix linter errors

maxbrunsfeld committed:
Fix propogation of included ranges for injections within injections

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

Commit:
Merge branch 'master' into tree-sitter-injections

maxbrunsfeld committed:
Preload main modules of language-html and language-javascript

maxbrunsfeld committed:
Fix error when injection points are added before grammars are loaded

maxbrunsfeld committed:
:arrow_up: tree-sitter

maxbrunsfeld committed:
Make select-larger-syntax-node command respect injected languages

Co-Authored-By: Ashi Krishnan <queerviolet@github.com>

Commit:
Folding works, though folding to a particular level doesn't respect nested grammars.

Co-Authored-By: Max <maxbrunsfeld@github.com>

Commit:
Spec for injected language folding.

Co-Authored-By: Max Brunsfeld <maxbrunsfeld@github.com>

maxbrunsfeld committed:
Only search for injection markers within a layer's included ranges

maxbrunsfeld committed:
WIP - put back parse queue

maxbrunsfeld committed:
Avoid O(injectionCount) linear searches in highlight iterator

maxbrunsfeld committed:
Avoid repeated characterIndexForPosition calls in HighlightIterator.seek

maxbrunsfeld committed:
:arrow_up: text-buffer

maxbrunsfeld committed:
Merge branch 'master' into tree-sitter-injections

maxbrunsfeld committed:
:shirt: Fix lint errors

maxbrunsfeld committed:
:arrow_up: tree-sitter

maxbrunsfeld committed:
Remove concurrency limit on parsing

maxbrunsfeld committed:
Represent the full node range set as null

We already had conditional logic based on whether the node range set
was 'full'.

maxbrunsfeld committed:
Rework handling of edited ranges

maxbrunsfeld committed:
Move helper functions to the bottom of the file

maxbrunsfeld committed:
Return containing tags from seek in the correct order

maxbrunsfeld committed:
Merge branch 'master' into tree-sitter-injections

maxbrunsfeld committed:
Silence complaint from standard 🙄

maxbrunsfeld committed:
Fix handling of folds inside highlighted tokens

maxbrunsfeld committed:
Incorporated injected languages in scope descriptors

maxbrunsfeld committed:
:arrow_up: language-html, language-javascript

fcomrqz commented:
Hi @maxbrunsfeld, 

I'm trying to use the `tree-sitter-vue`.

Syntax highlight is working well. 🎉 

Folding at given level and autocomplete only works for HTML.

In this pull request you leave a task incomplete:  
- [ ] Use injections for folding at a given nesting level.

Can you give me a hint about how can I fix this?
I don't know if is something related to the Vue parser, the Atom grammar, the `autocomplete-plus` or anything else.

In between I'm learning how to create a parser from .

Thanks. ✌️
